x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskManagementService.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.async;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.engine.DocumentMissingException;
+import org.elasticsearch.index.engine.VersionConflictEngineException;
+import org.elasticsearch.tasks.CancellableTask;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskAwareRequest;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.tasks.TaskManager;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class AsyncTaskManagementService<
```

Comment:
What is the purpose of this service? It looks to me like a simplified fork of what async search does when it comes to waiting for completion and storing the response. Yet I see only one user of these new abstractions. Are there going to be more users of this? Async search will not be migrated to it, right?

Reply:
We can migrate async search here in the future. The main difference is that this service relies on cascading cancellation and async search is preforming cancellation internally. So, some pieces of this logic are missing here.

Reply:
Cancellation is not a problem, we need to migrate async search to rely on cascading cancellation anyways. But the logic around storing partial responses and wait for completion is very different from what you have now.

Reply:
I think logic is pretty much the same but implementation is quite a bit different. Removing cancellation concerns allowed me to remove all the thread management and client calls from the task into the service and keep the task lightweight. I don't think a task should be a control structure or use clients. It should be just a simple light weight data transfer structure.

Reply:
One point where the behaviour differs at first glance is that in async search, we never want to return nor store a response before the initial onListShards listener callback. Hence we register a completion listener, but that is not invoked until onListShards is called.
What I am saying is that there are subtle differences that are independent from task cancellation, and I am not sure that async search can be migrated to this simplified version of async management service.